File: CircularLinkedList/CircularLinkedList.java
// Circlular Linked List ,....

class Node
{   
    int data;
    Node next;
}

public class CircularLinkedList  
{

    public void display()  {


  if ( head == null )   
		
     return ; 
		
  Node temp = head ; 
	
   do {
	
	System.out.println(temp.data);
	temp = temp.next ;
		 
   }while ( temp != head ) ;
	
  }

  Node head;

    public void addFirst(int val)
    {
        Node newNode = new Node();
        newNode.data = val;

        if(head == null)
        {
            newNode.next = newNode;
            head = newNode;
        }
        else
        {
             Node last = head;

             while(last.next != head)
             {
                 last = last.next;
             }

             last.next = newNode;
             newNode.next = head;
             head = newNode;
        }
    }
    
    
    public void addLast(int val)
    {
        Node newNode = new Node();
        newNode.data = val;

        if(head == null)
        {
            newNode.next = newNode;
            head = newNode;
        }
        else
        {
            Node last = head;

            while(last.next != head)
            {
                last = last.next;
            }

            last.next = newNode;
            newNode.next = head;
        }
    }
    
    public boolean search(int key)
    {
        if(head == null)
        
            return false;

        Node temp = head;

        do
        {
            if( temp.data == key)
                return true;
                
            temp = temp.next;

        } while(temp != head);

        return false;
    }
      
    public void delete(int key)
    {
        if(head == null)
            return;

        if(head.data == key && head.next == head)
        {
            head = null;
        }
        else if(head.data == key)
        {
            Node last = head;

            while(last.next != head)
            {
                last = last.next;
            }
            last.next = head.next;
            head = head.next;
        }
        else
        {
            Node current = head;

            while(current.next != head)
            {
                if(current.next.data == key)
                {
                    current.next = current.next.next;
                    break;
                }
                current = current.next;
            }
        }
    }

    public void print()
    {
        if(head == null)
            return;

        Node temp = head;

        do
        {
            System.out.println(temp.data);
            temp = temp.next;
        }while(temp != head);
    }

 public static void main(String args[])
    {

        CircularLinkedList list = new CircularLinkedList();
        list.addLast(10);
        list.addLast(20);
        list.addLast(30);
        list.addFirst(5);
        list.addLast(40);
        list.addLast(50);
        list.addLast(60);
        list.addLast(70);
        list.addLast(80);

        System.out.println("Circular Linked List: 10->20->30->10");

        int key = 20;
        System.out.println("Deleting Element: "+key);
        list.delete(key);

        System.out.println("After Deletion");
        list.print();
    
        System.out.println("Searching Element: "+20);
        if(list.search(20))
            System.out.println("Search Found");
        else
            System.out.println("Search Not Found");

        System.out.println("Searching Element: "+100);
        if(list.search(100))
            System.out.println("Search Found");
        else
            System.out.println("Search Not Found");
            
            
       }
          
}

File: JS_Algorithms/CircularLinkedList.js
/* Circular LinkedList */ 

var head = null;

class Node
{
    constructor(val)
     {
            this.data = val;
            this.next = null;
      }
}


function print()
{
    if(head == null)
    
        return;

    var temp = head;

    do
    {
        document.write(temp.data+"<br>");
        
        temp = temp.next;
        
    }while(temp != head);
}

function addFirst(val)
{
      var newNode = new Node(val);

      if(head == null)
      {
          newNode.next = newNode;
          head = newNode;
      }
      else
      {
           var last = head;

           while(last.next != head)
           
           {
               last = last.next;
           }

           last.next = newNode;
           newNode.next = head;
           head = newNode;
      }
}

function addLast(val)
{
      var newNode = new Node(val);

      if(head == null)
      
      {
          newNode.next = newNode;
          
          head = newNode;
      }
      else
      {
          var last = head;

          while(last.next != head)
          
          {
              last = last.next;
          }

          last.next = newNode;
          newNode.next = head;
      }
}

function search(key)
{
    if(head == null)
    
        return false;

    var temp = head;

    do
    {
        if(temp.data == key)
        
            return true;
            
        temp = temp.next;

    }while(temp != head);

    return false;
}

function del(key)
{
      if(head == null)
      
          return;

      if(head.data == key && head.next == head)
      {
          head = null;
      }
      else if(head.data == key)
      {
          var last = head;

          while(last.next != head)
          {
              last = last.next;
          }
          
          last.next = head.next;
          
          head = head.next;
      }
      else
      {
          var current = head;

          while(current.next != head)
          {
              if(current.next.data == key)
              {
                  current.next = current.next.next;
                  break;
              }
              
              current = current.next;
          }
      }
}

document.write("Inserting Element: "+10+"<br>");
addFirst(10);
document.write("Inserting Element: "+20+"<br>");
addFirst(20);
document.write("Inserting Element: "+30+"<br>");
addFirst(30);
document.write("Inserting Element: "+40+"<br>");
addFirst(40);

document.write("Circular Linked List Elements Are:<br>");

print();

document.write("Inserting Element: "+40+"<br>");
addLast(40);
document.write("Inserting Element: "+50+"<br>");
addLast(50);
document.write("Inserting Element: "+60+"<br>");
addLast(60);
document.write("Inserting Element: "+80+"<br>");
addLast(80);

document.write("Circular Linked List Elements Are:<br>");

print();

document.write("Circular Linked List: 40->30->20->10->40->50->60->80 <br>");

document.write("Searching Element: "+20+"<br>");
if(search(20))
    document.write("Search Found<br>");
else
    document.write("Search Not Found<br>");

document.write("Searching Element: "+100+"<br>");
if(search(100))
    document.write("Search Found<br>");
else
    document.write("Search Not Found<br>");
    
    
 var key = 20;
document.write("Deleting Element: "+key+"<br>");
del(key);

document.write("After Deletion<br>");

document.write("Printing Circular LnkedLisr again<br>");

print();

   




